components/autofill/src/db/schema.rs

```diff
@@ -173,20 +152,62 @@ fn upgrade(db: &Connection, from: i64) -> Result<()> {
             DROP TABLE credit_cards_data;
             ALTER TABLE new_credit_cards_data RENAME to credit_cards_data;
             ")?;
+        Ok(())
+    }
+}
+
+impl MigrationLogic for AutofillMigrationLogic {
+    const NAME: &'static str = "autofill db";
+    // AutofillDB has a slightly strange version history, so we start on v0.  See upgrade_from_v0()
+    // for more details.
+    const START_VERSION: u32 = 0;
+    const END_VERSION: u32 = 2;
+
+    fn prepare(&self, conn: &Connection) -> Result<()> {
+        AutofillMigrationLogic::define_functions(&conn)?;
+        conn.set_prepared_statement_cache_capacity(128);
+        Ok(())
+    }
+
+    fn init(&self, db: &Connection) -> Result<()> {
+        Ok(db.execute_batch(CREATE_SHARED_SCHEMA_SQL)?)
+    }
+
+    fn upgrade_from(&self, db: &Connection, version: u32) -> Result<()> {
```

Comment:
What's the reason for keeping the loop in sqlsupport? I'm wondering if this might deny the ability for some optimizations (eg, you could imagine a hypothetical scenario where it going from v5 to v7 could be done much more efficiently than going via v6.

Reply:
The main reason is that it made the upgrade function feel simpler if it always just handled one upgrade.  One thing weighing on my mind was the if statement in the autofill upgrade function, which made it so if you were on v0 then you didn't run the v1 upgrade code.  This ended up being fine, since the v0 upgrade deleted the table then we would recreate it at the end, but that was pure luck on my part.

I was thinking about the scenario where you wanted to skip versions.  We could make it so the upgrade function could return something to indicate that.  But the code was already getting complex enough, so I didn't implement it.

Reply:
FWIW, I get a strong YAGNI feeling on the idea of running multiple migrations at a time, especially since it would multiply the number of potential code paths to test. If we find ourselves wanting to do that at some point in the future, it seems like the sort of thing that we'll want to spend some careful thought and testing on.